tornado/locks.py

```diff
@@ -0,0 +1,62 @@
+__all__ = ['Condition']
+
+import collections
+
+from tornado import gen, ioloop
+from tornado.concurrent import Future
+
+
+class Condition(object):
+    """A condition allows one or more coroutines to wait until notified.
+
+    Like a standard Condition_, but does not need an underlying lock that
+    is acquired and released.
+
+    .. _Condition: http://docs.python.org/library/threading.html#threading.Condition
+    """
+
+    def __init__(self, io_loop=None):
+        self.io_loop = io_loop or ioloop.IOLoop.current()
+        self.waiters = collections.deque()  # Futures.
+
+    def __str__(self):
+        result = '<%s' % (self.__class__.__name__, )
+        if self.waiters:
+            result += ' waiters[%s]' % len(self.waiters)
+        return result + '>'
+
+    def wait(self, deadline=None):
+        """Wait for `.notify`. Returns a `.Future`.
+
+        Raises `.TimeoutError` if the condition is not notified before
+        ``timeout``, which may be specified in any form allowed by
+        `.IOLoop.add_timeout` (i.e. a `datetime.timedelta` or an absolute time
+        relative to `.IOLoop.time`)
+        """
+        waiter = Future()
+        self.waiters.append(waiter)
+        if deadline:
+            timed = gen.with_timeout(deadline, waiter, self.io_loop,
+                                     quiet_exceptions=gen.TimeoutError)
+
+            # Set waiter's exception after the deadline so notify(n) skips it.
+            gen.chain_future(timed, waiter)
+            return timed
+        else:
+            return waiter
+
+    def notify(self, n=1):
+        """Wake ``n`` waiters."""
+        waiters = []  # Waiters we plan to run right now.
+        while n and self.waiters:
+            waiter = self.waiters.popleft()
+            if not waiter.done():  # Might have timed out.
+                n -= 1
+                waiters.append(waiter)
+
+        for waiter in waiters:
+            waiter.set_result(None)
```

Comment:
Hmm. You're kind of responsible for that whether you want to be or not. With a separate loop like this, waiters woken by the nested notify will be woken up before the first notify has finished. It's not a big problem since you don't make any formal guarantees of FIFO ordering (and threaded Conditions are completely unordered), but it can be a surprise to have something that is *almost* FIFO. I think it's acceptable with or without the separate loop, but just wanted to make sure that if there is a reason for choosing one over the other that it is documented.

In a threaded Condition, the paired lock prevents this situation; a second notify can't happen until the first has finished. Do we need to reconsider pairing a lock with the Condition? It would be a lot more cumbersome to use (and much slower), but it would help adhere to the semantics of standard condition variables. On the other hand, it's so rarely necessary to use a lock on top of the single-threaded event loop and I don't think it adds that much value.